src/tsf/TfEditses.cpp

```diff
@@ -388,20 +388,20 @@ HRESULT CEditSessionObject::_GetTextAndAttribute(TfEditCookie ec, ITfRange* rang
             da.bAttr = TF_ATTR_INPUT;
 
             CompGuid.FillData(guidatom, ulcch0);
-            CompStr.Append(wstr0, ulcch0);
+            CompStr.append(wstr0, ulcch0);
         }
 
         textRange->Collapse(ec, TF_ANCHOR_END);
 
-        range.Release();
+        range.reset();
 
     } // out-most while for GUID_PROP_COMPOSING
 
 
     //
     // set GUID_PROP_CONIME_TRACKCOMPOSITION
     //
-    CComPtr<ITfProperty> PropertyTrackComposition;
+    wil::com_ptr_nothrow<ITfProperty> PropertyTrackComposition;

```

Comment:
I think the application of com_ptr_nothrow is correct, but I was surprised to see that all of these use the non-throwing version of wil's com_ptr since I assumed at least a few cases should throw an exception instead.

Was the new use verified to ensure that the nothrow variant matches the previous behavior in all cases?

Reply:
I took a look at atlcomcli.h and found `operator*` in `CComPtrBase` throws expection.
```
    T& operator*() const
    {
        ATLENSURE(p!=NULL);
        return *p;
    }
```
`#define ATLENSURE(expr) ATLENSURE_THROW(expr, E_FAIL)`

So here we can use `wil::com_ptr` instead of `wil::com_ptr_nothrow`?

Reply:
I would wait for one of the Terminal team members to respond before changing this.  But my intuition is that it was correct to use com_ptr_nothrow for the query cases, but the others should be com_ptr.

Reply:
Actually we were still discussing in the issue if `com_ptr_nothrow` is actually needed in TSF or not. I'd say that it isn't, exceptions can come from query functions but we actually need the "try" versions.

Reply:
> I took a look at atlcomcli.h and found operator* in CComPtrBase throws expection.

Yes but `wil`'s `com_ptr` does not. Besides, I don't think I've seen any uses of this particular operator.

Reply:
The `CEditSessionObject::_GetTextAndAttributeGapRange` function is already a little broken in that we were calling `std::wstring.append` here which I believe can technically throw in memory situations and we weren't catching it.

What I'd do here is wrap the entire function body in:
```
try
{
}
CATCH_RETURN();
return S_OK;
```

And just use the throwable versions of `wil::com_ptr` and continue to use the `std::wstring.append` inside of that.

That will ensure that any exception that might be thrown now or in future refactorings will be appropriately caught and `HRESULT`-ified for the COM caller.

Reply:
I added try blocks only for `CompGuid.insert` and `CompStr.append`.